File: jianzhioffer/firstround/jianzhi_27.py
# Definition for a binary tree node.
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution(object):
    def mirrorTree(self, root):
        """
        :type root: TreeNode
        :rtype: TreeNode
        """
        if not root:
            return
        stack = [root]
        while stack:
            node = stack.pop()
            if node.left: stack.append(node.left)
            if node.right: stack.append(node.right)
            node.left, node.right = node.right, node.left
        return root


if __name__ == '__main__':
    l1 = TreeNode(4)
    l2 = TreeNode(2)
    l3 = TreeNode(7)
    l4 = TreeNode(1)
    l5 = TreeNode(3)
    l6 = TreeNode(6)
    #l7 = TreeNode(9)

    l1.left = l2
    l1.right = l3
    l2.left = l4
    l2.right = l5
    l3.left = l6
    #l3.right = l7
    obj = Solution()
    node = obj.mirrorTree(l1)
    while node.left or node.right:
        print(node.left)
        print(node.right)
        node = node.left

File: jianzhioffer/secondround/jianzhi_27.cpp
#include <cstddef>
#include <vector>
using namespace std;

 //Definition for a binary tree node.
 struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 };


// //bfs,时间复杂度O(N)，空间复杂度O(N)
// class Solution {
// public:
//     TreeNode* mirrorTree(TreeNode* root) {
//         if(root==NULL) return NULL;
//         else if(root->left==NULL && root->right==NULL) return root;
//         vector<TreeNode*> node_list;
//         node_list.push_back(root);
//         while (!node_list.empty())
//         {
//             TreeNode* node = node_list.back();
//             node_list.pop_back();
//             if(node->left!=NULL) node_list.push_back(node->left);
//             if(node->right!=NULL) node_list.push_back(node->right);
//             swap(node->left, node->right);
//         }
//         return root;
//     }
// };

//递归,时间复杂度O(N)，空间复杂度O(N)
class Solution {
public:
    TreeNode* mirrorTree(TreeNode* root) {
        if(root==NULL) return NULL;
        TreeNode* tmp = root->left;
        root->left = mirrorTree(root->right);
        root->right = mirrorTree(tmp);
        return root;
    }
};

int main(int argc, char const *argv[])
{
    TreeNode t11 = TreeNode(4);
    TreeNode t12 = TreeNode(2);
    TreeNode t13 = TreeNode(7);
    TreeNode t14 = TreeNode(1);
    TreeNode t15 = TreeNode(3);
    TreeNode t16 = TreeNode(6);
    TreeNode t17 = TreeNode(9);

    t11.left = &t12;
    t11.right = &t13;
    t12.left = &t14;
    t12.right = &t15;
    t13.left = &t16;
    t13.right = &t17;

    Solution obj;
    TreeNode* result = obj.mirrorTree(&t11);
    return 0;
}